modules/aruco/src/aruco_detector.cpp

```diff
@@ -0,0 +1,1230 @@
+	}
+
+	/*
+	 * calculate the corner :: where the lines crosses to each other
+	 * clockwise direction		no clockwise direction
+	 *      0                           1
+	 *      .---. 1                     .---. 2
+	 *      |   |                       |   |
+	 *    3 .___.                     0 .___.
+	 *          2                           3
+	 */
+	for(int i=0; i < 4; i++){
+		if(inc<0)
+			nCorners[i] = _getCrossPoint(lines[ i ], lines[ (i+1)%4 ]);	// 01 12 23 30
+		else
+			nCorners[i] = _getCrossPoint(lines[ i ], lines[ (i+3)%4 ]);	// 30 01 12 23
+	}
+}
+
+static inline void findCornerInPyrImage(const float scale_init, const int closest_pyr_image_idx,
+                                        const std::vector<cv::Mat>& grey_pyramid, Mat corners,
+                                        const Ptr<DetectorParameters>& params) {
+    // scale them to the closest pyramid level
+    if (scale_init != 1.f)
+        corners *= scale_init; // scale_init * scale_pyr
+    for (int idx = closest_pyr_image_idx - 1; idx >= 0; --idx) {
+        // scale them to new pyramid level
+        corners *= 2.f; // *= scale_pyr;
+        // use larger win size for larger images
+        const int subpix_win_size = std::max(grey_pyramid[idx].cols, grey_pyramid[idx].rows) > 1080 ? 5 : 3;
+        cornerSubPix(grey_pyramid[idx], corners,
+                     Size(subpix_win_size, subpix_win_size),
+                     Size(-1, -1),
+                     TermCriteria(TermCriteria::MAX_ITER | TermCriteria::EPS,
+                                  params->cornerRefinementMaxIterations,
+                                  params->cornerRefinementMinAccuracy));
+    }
+}
+
+void ArucoDetector::detectMarkers(InputArray _image, OutputArrayOfArrays _corners, OutputArray _ids,
+                                  OutputArrayOfArrays _rejectedImgPoints) {
+    CV_Assert(!_image.empty());
+    CV_Assert(params->markerBorderBits > 0);
+    // check that the parameters are set correctly if Aruco3 is used
+    CV_Assert(!(params->useAruco3Detection == true &&
+                params->minSideLengthCanonicalImg == 0 &&
+                params->minMarkerLengthRatioOriginalImg == 0.0));
+
+    Mat grey;
+    _convertToGrey(_image.getMat(), grey);
+
+    // Aruco3 functionality is the extension of Aruco.
+    // The description can be found in:
+    // [1] Speeded up detection of squared fiducial markers, 2018, FJ Romera-Ramirez et al.
+    // if Aruco3 functionality if not wanted
+    // change some parameters to be sure to turn it off
+    if (!params->useAruco3Detection) {
+        params->minMarkerLengthRatioOriginalImg = 0.0;
+        params->minSideLengthCanonicalImg = 0;
+    }
+    else {
+        // always turn on corner refinement in case of Aruco3, due to upsampling
+        params->cornerRefinementMethod = CORNER_REFINE_SUBPIX;
+        // only CORNER_REFINE_SUBPIX implement correctly for useAruco3Detection
+        // TODO: update other CORNER_REFINE methods
```

Comment:
Todo

Reply:
fixed

Reply:
There are several todos for CORNER_REFINE_SUBPIX in the same file. Please take a look.